synthesizer/program/src/v2/bytes.rs

```diff
@@ -0,0 +1,199 @@
+                    None => return Err(error(format!("Mapping '{identifier}' is not defined"))),
+                },
+                ProgramDefinition::Struct => match self.structs.get(identifier) {
+                    Some(struct_) => {
+                        // Write the variant.
+                        1u8.write_le(&mut writer)?;
+                        // Write the struct.
+                        struct_.write_le(&mut writer)?;
+                    }
+                    None => return Err(error(format!("Struct '{identifier}' is not defined."))),
+                },
+                ProgramDefinition::Record => match self.records.get(identifier) {
+                    Some(record) => {
+                        // Write the variant.
+                        2u8.write_le(&mut writer)?;
+                        // Write the record.
+                        record.write_le(&mut writer)?;
+                    }
+                    None => return Err(error(format!("Record '{identifier}' is not defined."))),
+                },
+                ProgramDefinition::Closure => match self.closures.get(identifier) {
+                    Some(closure) => {
+                        // Write the variant.
+                        3u8.write_le(&mut writer)?;
+                        // Write the closure.
+                        closure.write_le(&mut writer)?;
+                    }
+                    None => return Err(error(format!("Closure '{identifier}' is not defined."))),
+                },
+                ProgramDefinition::Function => match self.functions.get(identifier) {
+                    Some(function) => {
+                        // Write the variant.
+                        4u8.write_le(&mut writer)?;
+                        // Write the function.
+                        function.write_le(&mut writer)?;
+                    }
+                    None => return Err(error(format!("Function '{identifier}' is not defined."))),
+                },
+            }
+        }
+        // Write the constructor.
+        match &self.constructor {
+            None => false.write_le(&mut writer)?,
+            Some(constructor) => {
+                // Write the variant.
+                true.write_le(&mut writer)?;
+                // Write the constructor.
+                constructor.write_le(&mut writer)?;
+            }
+        }
+        // Write the number of metadata entries and the metadata.
+        u16::try_from(self.metadata.len()).map_err(|e| error(e.to_string()))?.write_le(&mut writer)?;
+        for (_, metadata) in self.metadata.iter() {
+            // Write the metadata.
+            metadata.write_le(&mut writer)?;
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::Program;
+    use console::network::MainnetV0;
+
+    type CurrentNetwork = MainnetV0;
+
+    #[test]
+    fn test_bytes() -> Result<()> {
+        let program = r"
+program$2 token.aleo;
```

Comment:
From my humble POV, I'm not a fan of adding a new symbol `$` to the syntax - even if we don't typically expose aleo instructions to end users. 
- For the program version: what about a new keyword like `version` or `program_version` or `program_type` which we declare after `import` and before `program`?
- For metadata we could simply drop the `$`.

Reply:
I don't have strong opinions here. I'll leave it to the reviewers to arbitrate on syntax.
FWIW, I feel like the `version` keyword doesn't create enough of a conceptual separation but I can be convinced if I stare at some programs long enough.

Reply:
The frequency of introducing new program versions should also be taken into account. For instance, in Solidity versions are updated quite frequently. (Though the counterpart for solidity is Leo, and Aleo instructions is bytecode).

IMO program versions won't be updated frequently, so something more natural would be preferred.

Reply:
How integral is the addition of `$2`? Is the constructor or metadata be sufficient of a delineator for program version?

Program "version" in byte serialization is already hidden from the aleo instructions anyways.

Reply:
Constructors are optional. However, in a later PR, we do require that V2 programs contain `program_owner`, `edition`, and `updatable` metadata declarations. Perhaps that enough of a conceptual difference?

In general, `$2` is not integral. Since it sounds like the majority prefers to keep the syntax closer, I will attempt an alternate solution.

Reply:
I do like that this creates a clear indication to the program creator or users that the program is of a specific version instead of having to infer based on program syntax.